File: [DEPENDENCY]/oxmysql/lib/MySQL.ts
type Query = string | number;
type Params = Record<string, unknown> | unknown[];
type Callback<T> = (result: T | null) => void;

type Transaction = {
  query: Query;
} & ({ parameters?: Params } | { values?: Params });

const Wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
const assert = (condition: boolean, message: string) => {
  if (!condition) throw new TypeError(message);
};

const QueryStorage: string[] = [];

const safeArgs = (
  query: Transaction[] | string[] | Query,
  parameters?: Params,
  cb?: Callback<any>,
  transaction?: true
) => {
  if (typeof query === 'number') query = QueryStorage[query];

  if (transaction) assert(Array.isArray(query), `Transaction query expects an array, received ${typeof query}`);
  else assert(typeof query === 'string', `Query expects a string, received ${typeof query}`);

  if (parameters !== undefined)
    assert(
      typeof parameters === 'object' || Array.isArray(parameters),
      `Params expects object or array, received ${typeof parameters}`
    );

  if (cb !== undefined) assert(typeof cb === 'function', `Callback expects function, received ${typeof cb}`);

  return [query, parameters, cb];
};

function store(query: string) {
  assert(typeof query !== 'string', `Query expects a string, received ${typeof query}`);

  return QueryStorage.push(query);
}

function ready(callback: () => void) {
  setImmediate(() => {
    while (GetResourceState('oxmysql') !== 'started') Wait(50);
    callback();
  });
}

function query<T = any>(query: Query, parameters?: Params): Promise<T | null>;
function query<T = any>(query: Query, parameters?: Params, cb?: Callback<T>): void;
function query<T = any>(query: Query, parameters?: Params, cb?: Callback<T>): Promise<T | null> | void {
  if (cb) return exports.oxmysql.query(...safeArgs(query, parameters, cb));
  return exports.oxmysql.query_async(...safeArgs(query, parameters));
}

function single<T = Record<string | number, any>>(query: Query, parameters?: Params): Promise<T | null>;
function single<T = Record<string | number, any>>(query: Query, parameters?: Params, cb?: Callback<T>): void;
function single<T = Record<string | number, any>>(
  query: Query,
  parameters?: Params,
  cb?: Callback<T>
): Promise<T | null> | void {
  if (cb) return exports.oxmysql.single(...safeArgs(query, parameters, cb));
  return exports.oxmysql.single_async(...safeArgs(query, parameters));
}

function scalar<T = unknown>(query: Query, parameters?: Params): Promise<T | null>;
function scalar<T = unknown>(query: Query, parameters?: Params, cb?: Callback<T>): void;
function scalar<T = unknown>(query: Query, parameters?: Params, cb?: Callback<T>): Promise<T | null> | void {
  if (cb) return exports.oxmysql.scalar(...safeArgs(query, parameters, cb));
  return exports.oxmysql.scalar_async(...safeArgs(query, parameters));
}

function update(query: Query, parameters?: Params): Promise<number | null>;
function update(query: Query, parameters?: Params, cb?: Callback<number>): void;
function update(query: Query, parameters?: Params, cb?: Callback<number>): Promise<number | null> | void {
  if (cb) return exports.oxmysql.update(...safeArgs(query, parameters, cb));
  return exports.oxmysql.update_async(...safeArgs(query, parameters));
}

function insert(query: Query, parameters?: Params): Promise<number | null>;
function insert(query: Query, parameters?: Params, cb?: Callback<number>): void;
function insert(query: Query, parameters?: Params, cb?: Callback<number>): Promise<number | null> | void {
  if (cb) return exports.oxmysql.insert(...safeArgs(query, parameters, cb));
  return exports.oxmysql.insert_async(...safeArgs(query, parameters));
}

function prepare<T = unknown>(query: Query, parameters?: Params): Promise<T | null>;
function prepare<T = unknown>(query: Query, parameters?: Params, cb?: Callback<T>): void;
function prepare<T = unknown>(query: Query, parameters?: Params, cb?: Callback<T>): Promise<T | null> | void {
  if (cb) return exports.oxmysql.prepare(...safeArgs(query, parameters, cb));
  return exports.oxmysql.prepare_async(...safeArgs(query, parameters));
}

function transaction(query: Query, parameters?: Params): Promise<boolean>;
function transaction(query: Query, parameters?: Params, cb?: Callback<boolean>): void;
function transaction(query: Query, parameters?: Params, cb?: Callback<boolean>): Promise<boolean> | void {
  if (cb) return exports.oxmysql.transaction(...safeArgs(query, parameters, cb, true));
  return exports.oxmysql.transaction_async(...safeArgs(query, parameters, undefined, true));
}

export { store, query, single, scalar, update, insert, transaction, prepare, ready };
export default { store, query, single, scalar, update, insert, transaction, prepare, ready };

File: [DEPENDENCY]/oxmysql/lib/MySQL.lua
-- lib/MySQL.lua provides complete compatibility for resources designed for mysql-async
-- As of v2.0.0 this is the preferred method of interacting with oxmysql
-- * You can use mysql-async syntax or oxmysql syntax (refer to issue #77 or line 118)
-- * Using this lib provides minor improvements to performance and helps debug poor queries
-- * If using mysql-async syntax, a resource is not explicity bound to using oxmysql

-- todo: new annotations; need to see if I can get it working with metatables, otherwise it'll need stubs

local Store = {}

local function addStore(query, cb)
	assert(type(query) == 'string', 'The SQL Query must be a string')
	local store = #Store+1
	Store[store] = query
	if cb then cb(store) else return store end
end

local MySQL = {
	Sync = { store = addStore },
	Async = { store = addStore },

	ready = function(cb)
		CreateThread(function()
			repeat
				Wait(50)
			until GetResourceState('oxmysql') == 'started'
			cb()
		end)
	end
}

local type = type

local function safeArgs(query, parameters, cb, transaction)
	if type(query) == 'number' then query = Store[query] end

	if transaction then
		if type(query) ~= 'table' then
			error(("First argument expected table, received '%s'"):format(query))
		end
	else
		if type(query) ~= 'string' then
			error(("First argument expected string, received '%s'"):format(query))
		end
	end

	if parameters then
		local type = type(parameters)
		if type ~= 'table' and type ~= 'function' then
			error(("Second argument expected table or function, received '%s'"):format(parameters))
		end
	end

	if cb then
		local type = type(cb)
		if type ~= 'function' and (type == 'table' and not cb.__cfx_functionReference) then
			error(("Third argument expected function, received '%s'"):format(cb))
		end
	end

	return query, parameters, cb
end

local promise = promise
local oxmysql = exports.oxmysql
local Await = Citizen.Await
local GetCurrentResourceName = GetCurrentResourceName()

local function await(fn, query, parameters)
	local p = promise.new()
	fn(nil, query, parameters, function(result, error)
		if error then
			return p:reject(error)
		end

		p:resolve(result)
	end, GetCurrentResourceName, true)
	return Await(p)
end

setmetatable(MySQL, {
	__index = function(self, method)
		local state = GetResourceState('oxmysql')
		if state == 'started' or state == 'starting' then
			self[method] = setmetatable({}, {

				__call = function(_, query, parameters, cb)
					query, parameters, cb = safeArgs(query, parameters, cb, method == 'transaction')
					return oxmysql[method](nil, query, parameters, cb, GetCurrentResourceName, false)
				end,

				__index = function(_, index)
					assert(index == 'await', ('unable to index MySQL.%s.%s, expected .await'):format(method, index))
					self[method].await = function(query, parameters)
						return await(oxmysql[method], safeArgs(query, parameters, nil, method == 'transaction'))
					end
					return self[method].await
				end
			})

			return self[method]
		else
			error(('^1oxmysql resource state is %s - unable to trigger exports.oxmysql:%s^0'):format(state, method), 0)
		end
	end
})

local alias = {
	fetchAll = 'query',
	fetchScalar = 'scalar',
	fetchSingle = 'single',
	insert = 'insert',
	execute = 'update',
	transaction = 'transaction',
	prepare = 'prepare'
}

local alias_mt = {
	__index = function(self, key)
		if alias[key] then
			MySQL.Async[key] = MySQL[alias[key]]
			MySQL.Sync[key] = MySQL[alias[key]].await
			alias[key] = nil
			return self[key]
		end
	end
}

setmetatable(MySQL.Async, alias_mt)
setmetatable(MySQL.Sync, alias_mt)

_ENV.MySQL = MySQL

--[[
exports.oxmysql:query (previously exports.oxmysql:execute)
MySQL.Async.fetchAll = MySQL.query
MySQL.Sync.fetchAll = MySQL.query.await


exports.oxmysql:scalar
MySQL.Async.fetchScalar = MySQL.scalar
MySQL.Sync.fetchScalar = MySQL.scalar.await


exports.oxmysql:single
MySQL.Async.fetchSingle = MySQL.single
MySQL.Sync.fetchSingle = MySQL.single.await


exports.oxmysql:insert
MySQL.Async.insert = MySQL.insert
MySQL.Sync.insert = MySQL.insert.await


exports.oxmysql:update
MySQL.Async.execute = MySQL.update
MySQL.Sync.execute = MySQL.update.await


exports.oxmysql:transaction
MySQL.Async.transaction = MySQL.transaction
MySQL.Sync.transaction = MySQL.transaction.await


exports.oxmysql:prepare
MySQL.Async.prepare = MySQL.prepare
MySQL.Sync.prepare = MySQL.prepare.await
--]]
